parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,718 @@
+ * - Offsets: (size + 1) * 4-byte little-endian unsigned integers. `offsets[i]` represents the
+ * starting position of string i, counting starting from the address of `offsets[0]`. Strings
+ * must be stored contiguously, so we don’t need to store the string size, instead, we compute it
+ * with `offset[i + 1] - offset[i]`.
+ * - UTF-8 string data.
+ */
+public class VariantUtil {
+  public static final int BASIC_TYPE_BITS = 2;
+  public static final int BASIC_TYPE_MASK = 0x3;
+  public static final int PRIMITIVE_TYPE_MASK = 0x3F;
+  /** The inclusive maximum value of the type info value. It is the size limit of `SHORT_STR`. */
+  public static final int MAX_SHORT_STR_SIZE = 0x3F;
+
+  // The basic types
+
+  /**
+   * Primitive value.
+   * The type info value must be one of the values in the "Primitive" section below.
+   */
+  public static final int PRIMITIVE = 0;
+  /**
+   * Short string value.
+   * The type info value is the string size, which must be in `[0, MAX_SHORT_STR_SIZE]`.
+   * The string content bytes directly follow the header byte.
+   */
+  public static final int SHORT_STR = 1;
+  /**
+   * Object value.
+   * The content contains a size, a list of field ids, a list of field offsets, and
+   * the actual field values. The list of field ids has `size` ids, while the list of field offsets
+   * has `size + 1` offsets, where the last offset represents the total size of the field values
+   * data. The list of fields ids must be sorted by the field name in alphabetical order.
+   * Duplicate field names within one object are not allowed.
+   * 5 bits in the type info are used to specify the integer type of the object header. It is
+   * 0_b4_b3b2_b1b0 (MSB is 0), where:
+   *   - b4: the integer type of size. When it is 0/1, `size` is a little-endian 1/4-byte
+   *         unsigned integer.
+   *   - b3b2: the integer type of ids. When the 2 bits are 0/1/2, the id list contains
+   *           1/2/3-byte little-endian unsigned integers.
+   *   - b1b0: the integer type of offset. When the 2 bits are 0/1/2, the offset list contains
+   *           1/2/3-byte little-endian unsigned integers.
+   */
+  public static final int OBJECT = 2;
+  /**
+   * Array value.
+   * The content contains a size, a list of field offsets, and the actual element values.
+   * It is similar to an object without the id list. The length of the offset list
+   * is `size + 1`, where the last offset represent the total size of the element data.
+   * Its type info is: 000_b2_b1b0:
+   *   - b2: the type of size.
+   *   - b1b0: the integer type of offset.
+   */
+  public static final int ARRAY = 3;
+
+  // The primitive types
+
+  /** JSON Null value. Empty content. */
+  public static final int NULL = 0;
+  /** True value. Empty content. */
+  public static final int TRUE = 1;
+  /** False value. Empty content. */
+  public static final int FALSE = 2;
+  /** 1-byte little-endian signed integer. */
+  public static final int INT1 = 3;
```

Comment:
Minor: In the spec these are `INT8`, `INT16`, `INT32`, etc that match the Parquet physical types.